test/parallel_api/ranges/std_ranges_test.h

```diff
@@ -0,0 +1,455 @@
+    operator()(Policy, Algo algo, Args... args)
+    {
+        operator()(oneapi::dpl::execution::seq, algo, args...);
+        operator()(oneapi::dpl::execution::unseq, algo, args...);
+        operator()(oneapi::dpl::execution::par, algo, args...);
+        operator()(oneapi::dpl::execution::par_unseq, algo, args...);
+    }
+
+    template<typename Policy, typename Algo, typename Checker, typename Transform>
+    std::enable_if_t<!std::is_same_v<Policy, std::true_type> && Ranges == data_in>
+    operator()(Policy&& exec, Algo algo, Checker checker, Transform tr, auto... args)
+    {
+        constexpr int max_n = 10;
+        DataType data[max_n] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
+        DataType expected[max_n] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
+
+        auto expected_view = tr(std::ranges::subrange(expected, expected + max_n));
+        auto expected_res = checker(expected_view, args...);
+        {
+            Container cont(exec, std::ranges::begin(data), max_n);
+            typename Container::type& A = cont();
+
+            auto res = algo(exec, tr(A), args...);
+
+            //check result
+            if constexpr(RetTypeCheck)
+                static_assert(std::is_same_v<decltype(res), decltype(checker(tr(A), args...))>, "Wrong return type");
+
+            auto bres = ret_in_val(expected_res, expected_view.begin()) == ret_in_val(res, tr(A).begin());
+            EXPECT_TRUE(bres, (std::string("wrong return value from algo with ranges: ") + typeid(Algo).name() + 
+                typeid(decltype(tr(std::declval<Container&>()()))).name()).c_str());
+        }
+
+        //check result
+        EXPECT_EQ_N(expected, data, max_n, (std::string("wrong effect algo with ranges: ")
+            + typeid(Algo).name() + typeid(decltype(tr(std::declval<Container&>()()))).name()).c_str());
+    }
+    template<typename Policy, typename Algo, typename Checker, typename Transform>
+    std::enable_if_t<!std::is_same_v<Policy, std::true_type> && Ranges == data_in_out>
+    operator()(Policy&& exec, Algo algo, Checker checker, Transform tr, auto... args)
+    {
+        constexpr int max_n = 10;
+        DataType data_in[max_n] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
+        DataType data_out[max_n] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
+        DataType expected[max_n] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
+
+        auto src_view = tr(std::ranges::subrange(data_in, data_in + max_n));
+        auto expected_res = checker(src_view, expected, args...);
+        {
+            Container cont_in(exec, data_in, max_n);
+            Container cont_out(exec, data_out, max_n);
+
+            typename Container::type& A = cont_in();
+            typename Container::type& B = cont_out();
+
+            auto res = algo(exec, tr(A), B, args...);
+
+            //check result
+            if constexpr(RetTypeCheck)
+                static_assert(std::is_same_v<decltype(res), decltype(checker(tr(A), B, args...))>, "Wrong return type");
```

Comment:
good question,
I will try to remember (figure out)  why..

Reply:
`expected_res` has different type, because checker is launched with `expected_view`, not `tr(A)` as the tested `algo`.

Reply:
Can the expected data range be wrapped in a `Container` instead of a subrange to have matching return types?
This is not essential, and it does not block the PR.

Reply:
https://github.com/oneapi-src/oneDPL/issues/1777